File: chip_base.py
from exceptions import NotConnected, NoNamedPin, InvalidInput, InvalidOutput, InputFloating, DoesNotSupportBusInput

from collections import defaultdict

# Each output should put out either 0 or 1, but to support a bus type system where
# a input may be connected to the output of many nodes we support a constant value
# HIGH_Z which is equal to -1. If a not doesn't want to put out a values it can return
# this in order to say that it shouldn't be considered. On a bus only one node should
# output a value at any given time, and all others should be HIGH_Z. If this is not
# the case a error will be raised as this is generally a fail condition
HIGH_Z = -1

class ChipConnection:
    """ A struct like class for holding a node and output for connecting
    """
    def __init__(self, node, output):
        self.node = node
        self.output = output

    def __str__(self):
        return 'Connection : node={} output={}'.format(self.node, self.output)


class ChipBase:
    """ Base class that all IC subclass derive from. Handles
    all of the connection logic.
    """
    def __init__(self, name):
        self.name = name
        self.type = None
        self.discription = ""
        self.has_clock_input = False
        self.allow_bus_inputs = False

        self.pins = {}
        self.data = {}
        self.connections = defaultdict(list)

    def cook(self):
        """ Gather the values from all inputs and set them on
        all outputs.
        """
        raise NotImplementedError

    def check_connection(self, input):
        """ Check if a connection to be made is valid.

        Args:
            intput(str): The input on this node to check.
        """
        if not self.allow_bus_inputs and len(self.connections[input]) > 1:
            raise DoesNotSupportBusInput(('Node {} does not support bus input but has'
                ' multiple input connections for pin {}'.format(self.name, input)))

    def connect_node(self, node, output, inputs):
        """ Connect a upstream node output to this nodes input.

        Args:
            node (ChipBase): A node to connect.
            output (str): The output on the downstream node.
            inputs (str|list): The input on this node to connect or a list of inputs
        """

        if isinstance(inputs, list):
            for i in inputs:
                self.check_connection(i)
                self.connections[i].append(ChipConnection(node, output))
        else:
            self.check_connection(inputs)
            self.connections[inputs].append(ChipConnection(node, output))

    def get_input(self, input):
        """ Get the value from an upstream node.

        Args:
            input (int): An input index.
        
        Returns:
            bool: Value
        """
        data = self.connections.get(input)
        if not data:
            raise NotConnected('Input {} on Node {} has no connection'.format(input, self.name))
        for con in data:
            value = con.node.output_value(con.output)
            if not isinstance(value, int):
                raise InvalidOutput('Output {} on Node {} is not an int'.format(input, self.name))
            if value != HIGH_Z:
                return value
        raise InputFloating('Input {} on Node {} has connections but none are returning a vlue'.format(input, self.name))

    def output_value(self, output):
        """ Return the output value of a pin.

        Args:
            output (str): The output pin name
        
        Returns:
            int: The value
        """
        return self.data.get(output)

File: src/chip_base.cpp
#include "chip_base.h"

#include <exception>
#include <stdexcept>
#include <iostream>
#include <set>

using namespace IcSim;

ChipBase::ChipBase(std::string name)
{
    name_ = name;
}

/// Set a map of input pin names to pin numbers
void
ChipBase::set_pins(const PinNameMap& pins)
{
    pins_ = pins;
}

/// Change the internal state of this node, updating it's outputs from it's inputs
void
ChipBase::cook(bool& prop)
{
    // Do Nothing
    prop = false;
}

/// Cook and chips connected to this chip and then propagate to them
void
ChipBase::propagate()
{
    ChipSet prop_chips;
    // For each output node trigger a cook
    for (ChipBase* node : output_nodes_) {
        bool prop = false;
        node->cook(prop);
        if (prop) {
            // If that cook caused the node output values to change stage the node for propagation
            prop_chips.insert(node);
        }
    }
    // For all the nodes that should propagate, recurse and propagate them
    for (ChipBase* node : prop_chips) {
        node->propagate();
    }
}

/// Log all the input connections on this chip that will propagate and cause a cook
void
ChipBase::LogInputConnections() const
{
    ConnectionMap::const_iterator iter;
    for (iter = input_connections_.cbegin(); iter != input_connections_.cend(); iter++) {
        std::cerr << "InputConnections name: " << get_name();
        std::cerr << " input: " << iter->first;
        for (const ChipConnection& con : iter->second) {
            std::cerr << " output_name: " << con.node->get_name();
            std::cerr << " output_pin: " << con.pin;
        }
        std::cerr << std::endl;
    }
}

/// Log all the chips that should be propagated when this chip cooks.
void
ChipBase::LogOutputConnections() const
{
    for (ChipBase* node : output_nodes_) {
        std::cerr << "OutputConnections name: " << node->get_name() << " type: " <<  node->get_type() << std::endl;
    }
}

/// Check that the connection doesn't violate any rules like bus connections
void
ChipBase::check_connection(const int& input)
{
    if (!allow_bus_inputs()) {
        ConnectionMap::const_iterator iter = input_connections_.find(input);
        if (iter != input_connections_.cend()) {
            if (iter->second.size() > 1) {
                throw std::logic_error("Node "+name_+" does not support bus but has multiple connections.");
            }
        }
    }
}

/// Get a pin number for a named input.
int
ChipBase::get_pin_number(const std::string& pin_name) const
{
    PinNameMap::const_iterator iter = pins_.find(pin_name);
    if (iter != pins_.cend()) {
        return iter->second;
    }
    throw std::logic_error("Pin "+pin_name+" on Node "+name_+" not in PinNameMap.");
}

/// Connect the output of another chip to the input of this chip to form a connection
void
ChipBase::connect_input(ChipBase* node, const std::string& output, const std::string& input)
{
    connect_input(node, node->get_pin_number(output), get_pin_number(input));
}

/// Convenience method that takes a string pin name and a vector of inputs to connect.
void
ChipBase::connect_input(ChipBase* node, const std::string& output, const InputStringVector& inputs)
{
    for (const auto input: inputs) {
        connect_input(node, output, input);
    }
}

/// Connect the output of another chip to the input of this chip to form a connection
void
ChipBase::connect_input(ChipBase* node, const int& output, const int& input)
{
    // Check the connection is valid
    check_connection(input);
    // Make a holder for the connection
    ChipConnection input_holder;
    input_holder.node = node;
    input_holder.pin = output;
    // Insert into the map, this will make a new entry if required
    input_connections_[input].push_back(input_holder);

    // Connect the input on this node to an output on the upstream node so it can propagate.
    node->connect_output(this);
}

/// Convenience method that takes a vector of inputs to connect.
void
ChipBase::connect_input(ChipBase* node, const int& output, const InputVector& inputs)
{
    for (const auto input: inputs) {
        connect_input(node, output, input);
    }
}

/// Add a downsteam chip to this chip's set of output nodes for use when propagating.
void
ChipBase::connect_output(ChipBase* node)
{
    output_nodes_.insert(node);
}

/// Convenience method that takes a string pin name as input
const int
ChipBase::get_input(const std::string& input)
{
    return get_input(get_pin_number(input));
}

/// Get the value for an output on an upstream chip from an input on this chip
const int
ChipBase::get_input(const int& input)
{
    ConnectionMap::const_iterator iter = input_connections_.find(input);
    if (iter == input_connections_.cend()) {
        throw std::logic_error("Input "+std::to_string(input)+" on Node "+name_+" does not have existing connection.");
    }
    for (const ChipConnection& con : iter->second) {
        const int value = con.node->output_value(con.pin);
        if (value != HIGH_Z) {
            return value;
        }
    }
    throw std::logic_error("Input "+std::to_string(input)+" on Node "+name_+" has connections but none are returning a valid value.");
}

/// Convenience method that takes a string pin name as input
const int
ChipBase::output_value(const std::string& output) const
{
    return output_value(get_pin_number(output));
}

/// Output the current value in the PinDataMap for a given pin.
const int
ChipBase::output_value(const int& output) const
{
    PinDataMap::const_iterator iter = data_.find(output);
    if (iter != data_.cend()) {
        return iter->second;
    }
    throw std::logic_error("Output "+std::to_string(output)+" on Node "+name_+" does not contain a value.");
}


/// Convenience method that doesn't take propagate for use when setting inital values
/// in the chip constructor. Takes pin name as input.
void
ChipBase::set_data(std::string pin, const int& value)
{
    bool prop = false;
    set_data(pin, value, prop);
}

/// Convenience method that takes a string pin name as input
void
ChipBase::set_data(std::string pin, const int& value, bool& prop)
{
    set_data(get_pin_number(pin), value, prop);
}


/// Convenience method that doesn't take propagate for use when setting inital values
/// in the chip constructor.
void
ChipBase::set_data(int pin, const int& value)
{
    bool prop = false;
    set_data(pin, value, prop);
}

/// Set the data in the internal PinDataMap. Set the propagate reference if the data
/// is not the same as what was in the map previously.
void
ChipBase::set_data(int pin, const int& value, bool& prop)
{
    PinDataMap::iterator iter = data_.find(pin);
    if (iter == data_.end()) {
        data_[pin] = value;
    } else {
        int previous_value = iter->second;
        if (previous_value != value) {
            data_[pin] = value;
            // If the new value doesn't match the previous we should propagate
            prop = true;
        }
    }
}